netsize/netsize.go

```diff
@@ -0,0 +1,268 @@
+	for i := 0; i < bucketSize; i++ {
+		measurements[i] = []measurement{}
+	}
+
+	return &Estimator{
+		localID:      kbucket.ConvertPeerID(localID),
+		rt:           rt,
+		bucketSize:   bucketSize,
+		measurements: measurements,
+	}
+}
+
+// NormedDistance calculates the normed XOR distance of the given keys (from 0 to 1).
+func NormedDistance(p peer.ID, k ks.Key) float64 {
+	pKey := ks.XORKeySpace.Key([]byte(p))
+	ksDistance := new(big.Float).SetInt(pKey.Distance(k))
+	normedDist, _ := new(big.Float).Quo(ksDistance, keyspaceMaxFloat).Float64()
+	return normedDist
+}
+
+type measurement struct {
+	distance  float64
+	weight    float64
+	timestamp time.Time
+}
+
+// Track tracks the list of peers for the given key to incorporate in the next network size estimate.
+// key is expected **NOT** to be in the kademlia keyspace and peers is expected to be a sorted list of
+// the closest peers to the given key (the closest first).
+// This function expects peers to have the same length as the routing table bucket size. It also
+// strips old and limits the number of data points (favouring new).
+func (e *Estimator) Track(key string, peers []peer.ID) error {
+	e.measurementsLk.Lock()
+	defer e.measurementsLk.Unlock()
+
+	// sanity check
+	if len(peers) != e.bucketSize {
+		return ErrWrongNumOfPeers
+	}
+
+	logger.Debugw("Tracking peers for key", "key", key)
+
+	now := time.Now()
+
+	// invalidate cache
+	e.netSizeCache = nil
+
+	// Calculate weight for the peer distances.
+	weight := e.calcWeight(key, peers)
+
+	// Map given key to the Kademlia key space (hash it)
+	ksKey := ks.XORKeySpace.Key([]byte(key))
+
+	// the maximum age timestamp of the measurement data points
+	maxAgeTs := now.Add(-MaxMeasurementAge)
+
+	for i, p := range peers {
+		// Construct measurement struct
+		m := measurement{
+			distance:  NormedDistance(p, ksKey),
+			weight:    weight,
+			timestamp: now,
+		}
+
+		// keep track of this measurement
+		e.measurements[i] = append(e.measurements[i], m)
+
+		// find the smallest index of a measurement that is still in the allowed time window
+		// all measurements with a lower index should be discarded as they are too old
+		n := len(e.measurements[i])
+		idx := sort.Search(n, func(j int) bool {
+			return e.measurements[i][j].timestamp.After(maxAgeTs)
+		})
+
+		// if measurements are outside the allowed time window remove them.
+		// idx == n - there is no measurement in the allowed time window -> reset slice
+		// idx == 0 - the normal case where we only have valid entries
+		// idx != 0 - there is a mix of valid and obsolete entries
+		if idx == n {
+			e.measurements[i] = []measurement{}
+		} else if idx != 0 {
+			e.measurements[i] = e.measurements[i][idx:]
+		}
+
+		// if the number of data points exceed the max threshold, strip oldest measurement data points.
+		if len(e.measurements[i]) > MaxMeasurementsThreshold {
+			e.measurements[i] = e.measurements[i][len(e.measurements[i])-MaxMeasurementsThreshold:]
+		}
```

Comment:
The Go Compiler does not have any pointer aliasing analysis, it also has poor (/ next to none) interprocedural analysis for side effects of functions calls (and map accesses are runtime internal functions calls).
So it will conservatively optimize back to back loads from the map.

Also using `[i]` everywhere is confusing.

I would write this instead:
```suggestion
		measurements := append(e.measurements[i], m)

		// find the smallest index of a measurement that is still in the allowed time window
		// all measurements with a lower index should be discarded as they are too old
		n := len(measurements)
		idx := sort.Search(n, func(j int) bool {
			return measurements[j].timestamp.After(maxAgeTs)
		})

		// if measurements are outside the allowed time window remove them.
		// idx == n - there is no measurement in the allowed time window -> reset slice
		// idx == 0 - the normal case where we only have valid entries
		// idx != 0 - there is a mix of valid and obsolete entries
		if idx == n {
			del e.measurements[i] // REVIEW: let the map shrink again, there is a difference between a missing element and a zero element, the second one still take a bucket inside the map because you can do _, ok := m[k] and get ok == true
			return nil
		} // REVIEW: if idx != 0 do nothing, left slicing by 0 does nothing, and the act of reslicing is optimised to literally nothing (*if that can be proven inbound, if that is not then it's an if branch).
		measurements = measurements[idx:]

		// if the number of data points exceed the max threshold, strip oldest measurement data points.
		if len(measurements) > MaxMeasurementsThreshold {
			measurements = measurements[len(measurements)-MaxMeasurementsThreshold:]
		}
		
		e.measurements[i] = measurements
```

Reply:
`e.measurements` shouldn't grow or shrink. It should always have a constant number of keys of `bucketSize`.


Regarding your two `REVIEW` bits. We could actually get rid of the whole if statement and just write:

```go
measurements = measurements[idx:]
```

if `idx == n` this will return an empty slice. However, I believe if we just kept this line it would be a memory leak because we wouldn't free the underlying array. WDYT?